acme/api/middleware.go

```diff
@@ -0,0 +1,366 @@
+				"expected content-type to be in %s, but got %s", expected, ct)))
+			return
+		default:
+			expected := []string{"application/jose+json"}
+			if ct != "application/jose+json" {
+				api.WriteError(w, acme.MalformedErr(errors.Errorf(
+					"expected content-type to be in %s, but got %s", expected, ct)))
+				return
+			}
+			next(w, r)
+			return
+		}
+	}
+}
+
+// parseJWS is a middleware that parses a request body into a JSONWebSignature struct.
+func (h *Handler) parseJWS(next nextHTTP) nextHTTP {
+	return func(w http.ResponseWriter, r *http.Request) {
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			api.WriteError(w, acme.ServerInternalErr(errors.Wrap(err, "failed to read request body")))
+			return
+		}
+		jws, err := jose.ParseJWS(string(body))
+		if err != nil {
+			api.WriteError(w, acme.MalformedErr(errors.Wrap(err, "failed to parse JWS from request body")))
+			return
+		}
+		ctx := context.WithValue(r.Context(), jwsContextKey, jws)
+		next(w, r.WithContext(ctx))
+		return
+	}
+}
+
+// validateJWS checks the request body for to verify that it meets ACME
+// requirements for a JWS.
+//
+// The JWS MUST NOT have multiple signatures
+// The JWS Unencoded Payload Option [RFC7797] MUST NOT be used
+// The JWS Unprotected Header [RFC7515] MUST NOT be used
+// The JWS Payload MUST NOT be detached
+// The JWS Protected Header MUST include the following fields:
+//   * “alg” (Algorithm)
+//     * This field MUST NOT contain “none” or a Message Authentication Code
+//       (MAC) algorithm (e.g. one in which the algorithm registry description
+//       mentions MAC/HMAC).
+//   * “nonce” (defined in Section 6.5)
+//   * “url” (defined in Section 6.4)
+//   * Either “jwk” (JSON Web Key) or “kid” (Key ID) as specified below<Paste>
+func (h *Handler) validateJWS(next nextHTTP) nextHTTP {
+	return func(w http.ResponseWriter, r *http.Request) {
+		jws, ok := jwsFromContext(r)
+		if !ok || jws == nil {
+			api.WriteError(w, acme.ServerInternalErr(errors.Errorf("jws expected in request context")))
+			return
+		}
+		if len(jws.Signatures) == 0 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("request body does not contain a signature")))
+			return
+		}
+		if len(jws.Signatures) > 1 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("request body contains more than one signature")))
+			return
+		}
+
+		sig := jws.Signatures[0]
+		uh := sig.Unprotected
+		if len(uh.KeyID) > 0 ||
+			uh.JSONWebKey != nil ||
+			len(uh.Algorithm) > 0 ||
+			len(uh.Nonce) > 0 ||
+			len(uh.ExtraHeaders) > 0 {
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("unprotected header must not be used")))
+			return
+		}
+		hdr := sig.Protected
+		switch hdr.Algorithm {
+		case jose.RS256, jose.RS384, jose.RS512, jose.ES256, jose.ES384, jose.ES512, jose.EdDSA:
+			// we good
+		default:
+			api.WriteError(w, acme.MalformedErr(errors.Errorf("unsuitable algorithm: %s", hdr.Algorithm)))
+			return
+		}
```

Comment:
Should we enforce a minimum bit length for RSA keys?

Reply:
enforcing 2048. good spot.